dissect/volume/md/stream.py

```diff
@@ -0,0 +1,410 @@
+from __future__ import annotations
+
+import operator
+from typing import TYPE_CHECKING, NamedTuple, Optional
+
+from dissect.util.stream import AlignedStream, MappingStream
+
+from dissect.volume.exceptions import MDError
+from dissect.volume.md.c_md import SECTOR_SIZE, c_md
+
+if TYPE_CHECKING:
+    from dissect.volume.md.md import MD, Device
+
+
+class LinearStream(MappingStream):
+    """Implements a stream on a linear RAID set."""
+
+    def __init__(self, md: MD):
+        super().__init__()
+
+        offset = 0
+        for device in md.devices:
+            # Ignore special device roles for now
+            if device.raid_disk is None:
+                continue
+
+            device_size = device.data_size * SECTOR_SIZE
+            self.add(offset, device_size, device.open(), 0)
+            offset += device_size
+
+
+class Zone(NamedTuple):
+    zone_end: int
+    dev_start: int
+    devices: list[Device]
+
+
+class RAID0Stream(AlignedStream):
+    """Implements a stream on a RAID0 set."""
+
+    def __init__(self, md: MD):
+        self.md = md
+        self.devices = {dev.raid_disk: dev for dev in self.md.devices if dev.raid_disk is not None}
+        if len(self.devices) != md.raid_disks:
+            raise MDError(f"Missing disks in RAID0 set {self.md.uuid} ({self.md.name})")
+
+        # Determine how many strip zones we need to construct
+        # If a RAID0 set consists of devices with different sizes, additional strip zones
+        # may exist on the larger devices but not on the smaller ones
```

Comment:
If I understand correctly how `strip_zones` are defined, wouldn't a nicer algorithm be something along the lines of:

```
devices = sorted(self.devices.values(), key=operator.attrgetter("sectors"))
device_sizes = [device.sectors for device in devices]

next_device_size_offset = 0
next_device_size = device_sizes[next_device_size_offset]

zones = [Zone(next_device_size * len(devices), 0, devices)]

while True:
    prev_device_size_offset = next_device_size_offset
    prev_device_size = next_device_size

    next_device_size_offset = bisect.bisect(device_sizes, prev_device_size)

    if next_device_size_offset >= len(devices):
        break

    next_device_size = device_sizes[next_device_size_offset]

    strip_size = next_device_size - prev_device_size
    strip_devices = devices[next_device_size_offset:]

    zone_end = strip_size * len(strip_devices) + zones[-1].zone_end
    dev_start = ?? 
    zones.append(Zone(zone_end, dev_start, strip_devices))
```

Reply:
I tried to mimick the original algorithm as much as possible. With your suggestion you lose the importance of the `raid_disk` ordering, which is the index a device has in the RAID set. It's my understanding that the algorithm doesn't try to create the _optimal_ strip zones, but it tries to create strip zones within the device ordering of the RAID set.